pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/impl/FixedByteSingleValueMultiColumnReaderWriter.java

```diff
@@ -0,0 +1,207 @@
+    int chunkId = row / _numRowsPerChunk;
+
+    return _readers.get(chunkId).getInt(rowInChunk, column);
+  }
+
+  @Override
+  public long getLong(int row, int column) {
+    int rowInChunk = row % _numRowsPerChunk;
+    int chunkId = row / _numRowsPerChunk;
+
+    return _readers.get(chunkId).getLong(rowInChunk, column);
+  }
+
+  @Override
+  public float getFloat(int row, int column) {
+    int rowInChunk = row % _numRowsPerChunk;
+    int chunkId = row / _numRowsPerChunk;
+
+    return _readers.get(chunkId).getFloat(rowInChunk, column);
+  }
+
+  @Override
+  public double getDouble(int row, int column) {
+    int rowInChunk = row % _numRowsPerChunk;
+    int chunkId = row / _numRowsPerChunk;
+
+    return _readers.get(chunkId).getDouble(rowInChunk, column);
+  }
+
+  @Override
+  public String getString(int row, int column) {
+    int rowInChunk = row % _numRowsPerChunk;
+    int chunkId = row / _numRowsPerChunk;
+
+    return _readers.get(chunkId).getString(rowInChunk, column);
+  }
+
+  @Override
+  public void setInt(int row, int column, int value) {
+    ensureCapacity(row);
+    int rowInChunk = row % _numRowsPerChunk;
+    int chunkId = row / _numRowsPerChunk;
+    _writers.get(chunkId).setInt(rowInChunk, column, value);
+  }
+
+  @Override
+  public void setLong(int row, int column, long value) {
+    ensureCapacity(row);
+    int rowInChunk = row % _numRowsPerChunk;
+    int chunkId = row / _numRowsPerChunk;
+    _writers.get(chunkId).setLong(rowInChunk, column, value);
+  }
+
+  @Override
+  public void setFloat(int row, int column, float value) {
+    ensureCapacity(row);
+    int rowInChunk = row % _numRowsPerChunk;
+    int chunkId = row / _numRowsPerChunk;
+    _writers.get(chunkId).setFloat(rowInChunk, column, value);
+  }
+
+  @Override
+  public void setDouble(int row, int column, double value) {
+    ensureCapacity(row);
+    int rowInChunk = row % _numRowsPerChunk;
+    int chunkId = row / _numRowsPerChunk;
+    _writers.get(chunkId).setDouble(rowInChunk, column, value);
+  }
+
+  @Override
+  public void setString(int row, int column, String value) {
+    ensureCapacity(row);
+    int rowInChunk = row % _numRowsPerChunk;
+    int chunkId = row / _numRowsPerChunk;
+    _writers.get(chunkId).setString(rowInChunk, column, value);
+  }
+
+  @Override
+  public void close()
+      throws IOException {
+    _capacityInRows = 0;
+    _writers.clear();
+    _readers.clear();
+
+    for (PinotDataBuffer dataBuffer : _dataBuffers) {
+      dataBuffer.close();
+    }
+  }
+
+  /**
+   * Helper method to ensure capacity so that a row at specified location is available
+   * for write.
+   *
+   * @param row row id to be written at.
+   */
+  private void ensureCapacity(int row) {
+    if (row >= _capacityInRows) {
+
+      // Adding _chunkSizeInBytes in the numerator for rounding up. +1 because rows are 0-based index.
+      long buffersNeeded = (row + 1 - _capacityInRows + _numRowsPerChunk) / _numRowsPerChunk;
```

Comment:
shouldn't buffersNeeded simply increment by 1? whats the scenario where we would increment it by more than 1

Reply:
Yes, this supports adding row far into the future, without much penalty.

Reply:
should we allow that? I don't see a scenario for this and if this happens we should error out instead of adding buffers. At the least, set a max size and throw an error if we hit the max size.